src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
+            newNode.next = current;
+            if (current.prev != null) {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+            if (index == 0) {
+                head = newNode;
+            }
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return;
+        }
+
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size && index < EMPTY_LIST) {
+            throw new IndexOutOfBoundsException("Index is out of list");
+        }
+
+        Node<T> current = getNode(index);
+        return current.item;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size && index < EMPTY_LIST) {
+            throw new IndexOutOfBoundsException("Index is out of list");
+        }
+
+        Node<T> current = getNode(index);
+        current.item = value;
+        return current.item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size && index < EMPTY_LIST) {
+            throw new IndexOutOfBoundsException("Index is out of list");
+        }
+
+        Node<T> current = getNode(index);
+        unlinkNode(current);
+        size--;
+        return current.item;
+    }
+
+    private void unlinkNode(Node<T> node) {
+        node.item = null;
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.next = node.prev;
+        } else {
+            tail = node.prev;
+        }
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> current = getNode(object);
+        unlinkNode(current);
+        size--;
+        return true;
+    }
+
+    private Node<T> getNode(T value) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.item == value) {
+                break;
+            }
+            current = current.next;
+        }
+        return current;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
```

Comment:
Make `Node<T>` an inner class and place it at the end of the `MyLinkedList` class. Also, avoid using getters and setters in the inner class `Node` since we have access to private fields of the inner class.